warn/warn.go

```diff
@@ -960,6 +960,83 @@ func nativePackageWarning(f *build.File, fix bool) []*Finding {
 	return findings
 }
 
+// findEmptyReturns searches for return statements without a value and returns whether
+// the current list of statements terminates (either by a return or fail() statements on the current
+// level in all subranches.
+func findEmptyReturns(stmts []build.Expr, callback func(build.Expr)) bool {
+	if len(stmts) == 0 {
+		// May occur in empty else-clauses
+		return false
+	}
+	terminated := false
+	for _, stmt := range stmts {
+		switch stmt := stmt.(type) {
+		case *build.ReturnStmt:
+			if stmt.Result == nil {
+				callback(stmt)
+			}
+			terminated = true
+		case *build.CallExpr:
+			ident, ok := stmt.X.(*build.Ident)
+			if ok && ident.Name == "fail" {
+				terminated = true
+			}
+		case *build.ForStmt:
+			// Even if a for-loop is guaranteed to terminate in each iteration, buildifier still can't
+			// check whether the loop is not empty
+			findEmptyReturns(stmt.Body, callback)
```

Comment:
I don't understand the comment and why you call the function (but ignores its return value).

Reply:
The function tracks possible execution paths to determine whether it can reach the end where the implicit `return` is. The following example is considered as a termination point:

    if foo:
        return foo
    else:
        return bar
    # no implicit return here

however the following is not:

    for x in y:
        return x
    # implicit return here if the for loop above is empty